### PR TITLE
[WIP] Airy's Equation and Bessel's equation solving

### LINKED ISSUE #9819: ODE solver cannot solve Bessel equation
I tried to solve the Bessel equation with Sympy (version 0.7.6), but got a weird result (the Bessel function is expected):

```
>>> x = symbols('x')
>>> f = Function('f')
>>> eq=Derivative(f(x),x,2)*x*x+x*Derivative(f(x),x,1)+(x*x)*f(x)
>>> dsolve(eq)
f(x) == C1*(x**4/64 - x**2/4 + 1) + O(x**6)
```

### PR DESCRIPTION
**Pull request by nagarakshay**
Implementation of Airy's equation and Bessel's equation

Fixes #9819 
[1]Airy's equation

```
Eq(x ** 2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second kind respectively
```

[2]Bessel's equation

```
Eq((x+a)**2 * Derivative(y(x),x,x)+(x+a) * Derivative(y(x),x)+(λ**2* (x-a)**2-n ** 2)*y(x))
The above equation now has solution c1* besselj(n,λ*x - λ*a)+c2 *bessely(n,λ*x - λ*a)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second kind respectively
```

### CODE REVIEW

**Comment by nagarakshay** [comment]
@zanzibar7 @cbm755 Ping!

File: sympy/solvers/ode.py [hunk 1]
```diff
@@ -303,7 +303,9 @@
     "nth_linear_constant_coeff_variation_of_parameters",
     "nth_linear_euler_eq_nonhomogeneous_variation_of_parameters",
     "Liouville",
+    "2nd_linear_airy",
     "2nd_power_series_ordinary",
+    "Bessel_Equation",
```
Comment (zanzibar7) [inline sympy/solvers/ode.py]: Better to name new hints consistently with existing names.  I suggest "2nd_linear_bessel"

File: sympy/solvers/ode.py [hunk 2]
```diff
@@ -3780,6 +3825,17 @@ def ode_2nd_power_series_regular(eq, func, order, match):
             return Eq(f(x), collect(finalseries1 + finalseries2,
                 [C0, C1]) + Order(x**terms))
 
+def ode_Bessel_Equation(eq, func, order, match):
```
Comment (zanzibar7) [inline sympy/solvers/ode.py]: Suggest `ode_2nd_linear_bessel` .

**Comment by zanzibar7** [comment]
Looks like the tests failed.  Before submitting a pull request, run the test suite locally and make sure things pass at your end with `bin/test` and `bin/doctest`.

We also need to enhance `classify_ode` under `order == 2` to recognize Airy and Bessel equations, so that these get called correctly.

And we need tests in `test_ode.py` to show 

* Equations are appropriately classified
* Equations are solved correctly

**Comment by nagarakshay** [comment]
@zanzibar7 Ok now the tests pass from my side. I have amended the name of equation too and added the tests too. Is there anything else?

File: sympy/solvers/ode.py [hunk 3]
```diff
@@ -1202,8 +1203,25 @@ class in it.  Note that a hint may do this anyway if
                             not check.has(zoo) and not check.has(-oo):
                             coeff_dict = {'p': p, 'q': q, 'x0': point, 'terms': terms}
                             matching_hints["2nd_power_series_regular"] = coeff_dict
-
-
+                            #If the ODE has regular singular point at x0 and is of the form
+                            #Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x) thus Bessel's equation
+                            if p==1:
+                                b4 = Wild('b4', exclude=[x,f(x),df])
```
Comment (zanzibar7) [inline sympy/solvers/ode.py]: Does b4 have to be an integer, or will this work with fractional Bessel orders?  

Also, can we catch equations that are not already in standard form?

Comment (nagarakshay) [inline sympy/solvers/ode.py]: Yes b4 can be anything. No, please elaborate what sort of equations should the Bessel's equation support?

**Comment by jksuom** [comment]
> what sort of equations should the Bessel's equation method of solving support?

It seems that the class of equations whose solutions can be expressed in terms of Bessel functions is large. To take them all would be quite a big task. But I think that at least those equations could be included that have solutions of the type ``J(2*x)`` or ``J(x + 1)`` where ``J`` is a Bessel function. That is, equations obtained from Bessel's equation by a simple change of variable.

**Comment by nagarakshay** [comment]
Sure, I will add it right away.

**Comment by nagarakshay** [comment]
@jksuom @zanzibar7 Does this look ok?

**Comment by jksuom** [comment]
I think that one should start by looking for the singular points. If there is only one, then that could be used as ``x0``.

**Comment by nagarakshay** [comment]
As you say, I started to implemented the idea but there was a small hindrance.
Here is what happened effectively.
 I passed the equation 
>>eq = (x-2)**2*f(x).diff(x, 2)+ (x-2)*(f(x).diff(x)) + 25*(x-2)**2*f(x)
>>a3 = Wild('a3', exclude=[f(x), df, f(x).diff(x, 2)])
    b3 = Wild('b3', exclude=[f(x), df, f(x).diff(x, 2)])
    c3 = Wild('c3', exclude=[f(x), df, f(x).diff(x, 2)])
>>reduced_exp
25*x**2*f(x) + x**2*Derivative(f(x), x, x) - 100*x*f(x) + x*Derivative(f(x), x) - 4*x*Derivative(f(x), x, x) + 100*f(x) - 2*Derivative(f(x), x) + 4*Derivative(f(x), x x)
>>deq = b3*df + a3*(f(x).diff(x, 2)) + c3*f(x)
>>r = collect(reduced_eq,[f(x).diff(x, 2), f(x).diff(x), f(x)]).match(deq)
None
I cant understand why is Null being returned?

**Comment by nagarakshay** [comment]
Ping @jksuom

**Comment by jksuom** [comment]
It seems to me that None was returned because ``reduced_eq`` is not of the type expected by ``deq``.

**Comment by nagarakshay** [comment]
@jksuom I didn't get you. After collect(reduced_eq,[f(x).diff(x, 2), f(x).diff(x), f(x)]
 the expression should be of the form required by deq.

**Comment by jksuom** [comment]
That was only a guess since it is hard to see what the actual values are. The code should be surrounded by triple backticks to preserve the original form.

**Comment by nagarakshay** [comment]
I have a question. Do brackets matter while matching two expression? Or in other words if collect returns a parenthesizedpolynomial will a3 match it?

**Comment by jksuom** [comment]
I do not know how match behaves. I suspect that match does not see brackets or parentheses. But you should experiment and investigate that.

**Comment by jksuom** [comment]
It seems that match is not able to deal with more complicated expressions, because it does internally expand them even if the coefficients have been collected into a parenthesized expression. Probably you should try to find the coefficients by some other method. Perhaps by analysing the arguments of the collected equation in the code.

**Comment by nagarakshay** [comment]
I matched all of the Derivatives individually and now it works perfectly also i amended the regular power series code to make it work with arbitary singular points (it previously had no mechanism for that)but some other tests are failing due to which this PR is getting delayed. I will try to address them as soon as possible.

**Comment by nagarakshay** [comment]
@jksuom Is there a way of converting an object of class Derivative to an object of class Add?
Actually when the function has only a single Derivative
```
>>>reduced_exp=Derivative(f(x),x,2)
>>>reduced_exp.args #This should only return the current derivative
[f(x),x,x]

**Comment by jksuom** [comment]
You could probably use ``Add.make_args(reduced_exp)``.

**Comment by nagarakshay** [comment]
@jksuom Yeah that works perfectly. Thanks

**Comment by asmeurer** [comment]
What is the status of this?

**Comment by nagarakshay** [comment]
@asmeurer Sorry for the delay. I was busy with my university exams and end term assessments. I have almost completed this PR 2-3 test cases were failing. I will update in 2 days max.

**Comment by RituRajSingh878** [comment]
what's the status of this pr .can I start from the beginning by new pr.

**Comment by oscarbenjamin** [comment]
I would think that it is worth starting from this PR if you want to work on the same thing.

**Pull request closed without merging**